src/main/java/org/swisspush/reststorage/FilePutter.java

```diff
@@ -0,0 +1,162 @@
+package org.swisspush.reststorage;
+
+import io.vertx.core.Handler;
+import io.vertx.core.file.AsyncFile;
+import io.vertx.core.file.CopyOptions;
+import io.vertx.core.file.FileSystem;
+import io.vertx.core.file.OpenOptions;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.swisspush.reststorage.util.LockMode;
+
+import java.io.File;
+import java.util.UUID;
+
+
+/**
+ * Used by {@link FileSystemStorage} to execute file PUTs.
+ */
+public class FilePutter {
+
+    public static interface FilePutterCallbacks {
+        /** Same as {@link FileSystemStorage#delete(String, String, LockMode, long, boolean, boolean, Handler)}. */
+        void delete(String path, String lockOwner, LockMode lockMode, long lockExpire, boolean confirmCollectionDelete, boolean deleteRecursive, final Handler<Resource> handler);
+        /** Same as {@link FileSystemStorage#canonicalize(String)}. */
+        String canonicalize(String path);
+    }
+
+    private static final int MOVE_TO_FINAL_DESTINATION_ATTEMPT_LIMIT = 256;
+    private static final Logger log = LoggerFactory.getLogger(FilePutter.class);
+    private final CopyOptions moveOptions = new CopyOptions().setReplaceExisting(true);
+    private final FileSystem fileSystem;
+    private final String fullPath;
+    private final FilePutterCallbacks callbacks;
+    private final Handler<Resource> onCompleteHandler;
+    private String tmpFilePath;
+    private String tmpFilePathAbs;
+    private String tmpFileParentPath;
+    private volatile boolean executed = false;
+    private int moveToFinalDestinationAttemptCount = 0;
+
+    /**
+     * Package-private because currently only used internally.
+     */
+    FilePutter(FileSystem fileSystem, String fullPath, FilePutterCallbacks callbacks, Handler<Resource> onCompleteHandler) {
+        this.fileSystem = fileSystem;
+        this.fullPath = fullPath;
+        this.callbacks = callbacks;
+        this.onCompleteHandler = onCompleteHandler;
+    }
+
+    /**
+     * <p>Triggers the configured task. This method should only get called once per
+     * instance!</p>
+     *
+     * @throws IllegalStateException
+     *      Eg. in case method gets called more than one time.
+     */
+    public synchronized void execute() {
+        if( executed ){
+            throw new IllegalStateException( "This putter already got executed." );
+        }
+        this.executed = true;
+        // Setup required context.
+        this.tmpFilePath = "/.tmp/uploads/" + new File(fullPath).getName() + "-" + UUID.randomUUID().toString() + ".part";
```

Comment:
The "fullPath" seems not to be canonicalized anywhere - but it probably should....